_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+- **S**hared - współdzielony pomiędzy wiele komponentów i serwisów.
+- **H**ydrated - trwały i z możliwością ponownego zasilenia z zewnętrznego źródła jak np. local storage.
+- **A**vailable - dostępny cały czas, niezależnie od sposobu nawigacji po aplikacji, np. podczas przechodzenia i cofania się w aplikacji prezentującej złożony wniosek.
+- **R**etrieved - zdolny do przechowywania danych pochodzących z zewnętrznych źródeł, co pozwala na przykład zapisać wynik żądania HTTP i nie wykonywać go po raz kolejny w celu otrzymania tych samych informacji.
+- **I**mpacted - zdolny do zmieniania się pod wpływem akcji wykonywanych przez komponenty i serwisy.
+
+## Architektura Redux
+No dobrze, to jeśli już wiemy, że w naszym dobrym interesie leży skorzystanie z dobrodziejstw architektury Redux, to spójrzmy, co tak naprawdę przez to rozumiemy.
+
+![Redux](/assets/img/posts/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej/redux.png)
+<span class="img-legend">Architektura Redux z lotu ptaka. Źródło: dokumentacja NgRx</span>
+
+Na razie powyższy diagram może wydawać się nieco tajemniczy, zatem już spieszę z wyjaśnieniami.
+
+* Jak wspomniałem we wstępie, Redux zakłada istnienie globalnego, niemutowalnego bytu, przechowującego stan aplikacji - **store**. Fizycznie jest to obiekt w formacie JSON, którego struktura definiowana jest przez programistę. 
+* W celu wywołania zmian w store komponenty aplikacji muszą wywoływać **akcje** (ang. _actions_). Reprezentują one konkretne zdarzenia zachodzące w systemie i niosą ze sobą konkretne informacje.
+* Akcje są przechwytywane przez **reducery** (ang. _reducers_). Reducerem nazywamy czystą funkcję (ang. _pure function_), która konsumuje akcję i, w zależności od jej przeznaczenia oraz zgodnie z logiką reducera, zwraca odpowiednio zmodyfikowany, nowy store. 
+* Dane ze store do komponentów trafiają poprzez **selektory** (ang. _selectors_). Wszystkie zmiany stanu aplikacji są propagowane do reagujących na nie selektorów dzięki mechanizmowi Observable znanym z biblioteki RxJs, o której więcej można przeczytać [w tym wpisie]({% post_url pl/2020-01-09-rxjs-introduction %}). Selektory powinny otrzymywać wyłącznie dane potrzebne do działania komponentów, w których są używane.
+* W przypadkach, kiedy wywołanie akcji powinno pociągnąć za sobą dowolne działanie niezwiązane bezpośrednio z aplikacją (np. zapytanie do bazy danych, czy żądanie do zewnętrznej usługi) - do gry wchodzą **efekty** (ang. _effects_). Podobnie jak reducery potrafią one reagować na konkretne akcje i wykonać przypisane im zadanie. Mogą one również wywołać kolejną akcję, która trafi do reducera, a co za tym idzie - do store. Wywołania efektów mogą być zarówno synchroniczne, jak i asynchroniczne.
```

Comment:
efekty to jest ngrxowy byt/nazewnictwo. bo przecież dla asynchronicznych operacji możemy mówić o thunk/saga/... czy nawet ogólnie o middleware

Reply:
True. Spróbuję to uogólnić do punktu o middleware właśnie